readme.txt
Ultima 9 Tool Suite
Copyright (C) 2000 David Churchill
Copyright (C) 2014 Christopher Thielen

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

The original author can be contacted by electronic mail at froods@alphalink.com.au
The original version can be found at http://home.alphalink.com.au/~froods/

The current author can be contacted via GitHub.
GitHub contact information and the latest version can be found at https://github.com/cthielen/u9tools/

Ultima 9 is probably a registered trademark of Origin Systems, Inc.
The author of this program is in no way affiliated with Origin Systems.


The unconverted tools in this package use the funky Allegro graphics library available at
http://www.talula.demon.co.uk/allegro/ and were compiled with DJGPP, a free compiler for the x86 platform, get it at http://www.delorie.com/

The updated tools use SDL (libsdl.org) and were compiled using the GNU GCC compiler.

---------
Credits
---------
The format of the type 1 encoding in the sound files was determined by Josef Drexler.
Many thanks to him as we can now extract the wonderful music from this game.  Get
Josef's U9Decode program at http://publish.uwo.ca/~jdrexler/ultima/



                         -------------------------------
                          Message from Original Release
                         -------------------------------

Hi,
   This is the first release of a set of tools to extract various pieces of 
data from the Ultima 9 data files.  Not much is completed at this stage, and
functionality is minimal.  I mostly wrote the tools to confirm my guesses as
to the data formats.  What I've discovered so far can be found in the *format.txt
files that should have come with this package.  I only have access to a DOS
compiler (DJGPP), and don't have a great deal of experience programming in C
(most of my experience is in Pascal).  If someone wants to take the information
in these text files and produce a browser of some kind that would be great.

If you've got DJGPP and Allegro installed, you should be able to make everything
by running mkall.bat (yup, still need to read up on makefiles too...), but fear
not, precompiled exe's are included.  Note that this is all distributed under
the GPL (see COPYING), and that there is no warranty - that said, it's all 
worked fine for me ;)

Please don't hesitate to send any comments to froods@alphalink.com.au

Have fun,
		Dave.




Note: Most of these programs perform little to no error checking - and will
      overwrite files that already exist without warning.  It would probably
      be wise to copy the data file you're working on to a new directory, and
      make sure you run the program from that directory.


What's Included
---------------

u9xt.exe <filename.flx> <id number> <output file>

The flx files in static and sound are just lots of little files all globbed into
one.  This program will extract the id'th (zero based) item from filename.flx
and dump it info output file.


u9books.exe <bookfile.flx>

The books file (eg. books-en.flx) contains all the books in the game, as well
as the shops (ooh, I see some cheap goods in the near future), and journal entries
for major items.  Running this will create a u9-books directory under the current
directory and dump a heap of html files in it.  Once it's finished (takes a little
time, be patient), simply open up index.html.  Note that the formatting codes
in the books file aren't translated very well yet...


u9snd.exe [-i] [-r <min>-<max>] -f <inputfile>

This program will extract all sounds except speech (type 2 encoded) from the sound
files (music.flx, speech.flx, sfx.flx).  Use -i to make the program only create an
index, -r to set the range of items to work on (get id numbers from the index), -f
to set the filename.  Only -f is required, by default the program will extract 
_all_ sounds in a particualar file.  Be sure you have a heap of space free!


u9txv.exe <texturefilename>

This program will display the textures in a texture file (only texture8.* at the
moment).  It requires 1024x768x16 to run, although this is easily changed if
you can recompile it.  Use the '<' and '>' keys (well, actually ',' and '.') to
move to new textures - it's probably better to use '>', as it currently displays
the textures sequentially across the screen.  To exit simply hit escape.


u9bmp.exe [gfx-mode-id]

This program will display the images contained in bitmap16.flx, which must be
in the current directory.  gfx-mode-id is optional; to get a list of valid
values run u9bmp with -? as a parameter.  A video card capable of a 16-bit 
resolution is needed for this.  Yours probably is, but might still not work.
If you have problems try installing a VESA BIOS driver (like the Scitech 
Display Doctor), otherwise contact me and I see if I can help out.
NB: The function to export the images to .bmp files creates 16-bit colour
    bitmap files.  AFAIK the format it produces conforms to the standard,
    and at the least mspaint will read them.  If you have trouble loading
    them into your favourite image manipulation program, try loading them
    in Paint and then saving them from there - better yet, try to fix the
    bug in my .bmp code :P


Use the following keys to browse the images:

'v' - foward 1       'f' - back 1
'c' - foward 10      'd' - back 10
'x' - foward 100     's' - back 100
'z' - foward 1000    'a' - back 1000

<cr> - dump current image(s) to .bmp files (see note above).

' ' (space) - toggle mask (removes transparency info, try this if an image looks
              whacked).
esc - exit




Other Notes of Interest:
------------------------

The texture* files in static only relate to particular maps.

Each map has a terrain, fixed, and nonfixed, all with common extensions.

There are a /heap/ of test maps!

The terrain appears to simply be a height map.

To achieve caves, rock arches you can walk under - objects are used (this can
easily be seen by setting object distance to near, terrain distance to far).

The bitmap* files have coresponding sdinfo* files, there is a bitmap-
sdinfo pair for each type of texture: 8-bit (sh), 16-bit (16), and
compressed (C).

text.flx and *brk* contain the conversation and other text, text.dat seems to 
be an index to these (try setting your language to the extension of each of the
*brk* files in turn, in options.ini!)

yiqccube.dat is probably a colour cube for a colour compression scheme used
at least on 3dfx cards.

The file 00000409.256 is a 256 colour bitmap in windows format - simply change
the extension to .bmp


ChangeLog
2014.04.26

* Ported u9bmp (as u9bmp-linux) to SDL. Used in debugging a Wine bug:
  http://bugs.winehq.org/show_bug.cgi?id=13683

2000.01.26 - r4:

* Added support for type 1 encoding, and a few other features, to u9snd.exe.  This
  means that the music and the rest of the sound effects can be extracted (see 
  "Credits" and "What's Included").


2000.01.19 - r3:

* Added export feature to u9bmp.exe, hit enter to dump to current image(s) to
  .bmp files


2000.01.15 - r2:

* Added u9bmp.*, to view the images in the bitmap files (see "What's Included"
  for further information).



flxx.c
/*

Ultima 9 eXTractor Functions
Copyright (C) 2000 David Churchill

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

The author can be contacted by electronic mail at froods@alphalink.com.au
The most recent version can be found at http://home.alphalink.com.au/~froods/

Ultima 9 is probably a registered trademark of Origin Systems, Inc.
The author of this program is in no way affiliated with Origin Systems.


   FLX eXtractor for globbed files that come with UIX,
   and probably earlier Ultimas as well.

   Currently supports three functions:

   * Opening a FLX file
   * Retrieving the number of contained files
   * Extracting a file

   The extraction function simply moves the file pointer
   to the right place, and returns the length of the file.

   *** WARNNG ***
   Only /minimal/ error checking is done.

*/


#include "flxx.h"


FILE *flxx_open(char *fname)
{
   if ((fname == NULL) || (strlen(fname) == 0))
      return NULL;

   return fopen(fname, "rb");
}



int flxx_numitems(FILE *fp)
{
   int numitems = -1;


   if (fp == NULL)
      return -1;

   fseek(fp, 0x50, SEEK_SET);
   fread(&numitems, 4, 1, fp);
   return numitems;
}



int flxx_getitem(FILE *fp, int id)
{
   struct flxx_dirent { int ofs, len; } de;

   if ((fp == NULL) || (id >= flxx_numitems(fp)))
      return -1;

	fprintf(stdout, "flxx_getitem: requested id %d\n", id);

   fseek(fp, 0x80 + 8*id, SEEK_SET);
   fprintf(stdout, "flxx_getitem: flxx seek to %d (reading header?)\n", 0x80 + 8*id);
   if (fread(&de, sizeof(struct flxx_dirent), 1, fp) != 1)
      return -1;

   fprintf(stdout, "flxx_getitem: flxx seek to %d (0x%x) (final seek)\n", de.ofs, de.ofs);
   fseek(fp, de.ofs, SEEK_SET);

   return de.len;
}






u9xt.c
/*

Ultima 9 eXTractor
Copyright (C) 2000 David Churchill

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

The author can be contacted by electronic mail at froods@alphalink.com.au
The most recent version can be found at http://home.alphalink.com.au/~froods/

Ultima 9 is probably a registered trademark of Origin Systems, Inc.
The author of this program is in no way affiliated with Origin Systems.


    Usage: uxt <filename.flx> <id number> <output filename>

*/


#include <stdio.h>
#include <stdlib.h>
#include "flxx.h"

#define BUF_SIZE 32768


int main(int argc, char *argv[])
{
   FILE *fp, *fout;
   int len;
   char buf[BUF_SIZE];


   if (argc != 4)
   {
      fprintf(stderr,
              "usage: %s <filename.flx> <id number> <output filename>\n",
              argv[0]);
      exit(-1);
   }

   fp = flxx_open(argv[1]);
   if (fp == NULL)
   {
      fprintf(stderr, "error opening %s\n", argv[1]);
      exit(-1);
   }

   printf("extracting item %i (max %i)\n",
           atoi(argv[2]), flxx_numitems(fp)-1);

   len = flxx_getitem(fp, atoi(argv[2]));
   if (len < 0)
   {
      fprintf(stderr, "error extracting id %i, possibly non-existant\n",
              atoi(argv[2]));
      exit(-1);
   }


   fout = fopen(argv[3], "wb");
   if (fout == NULL)
   {
      fprintf(stderr, "error writing %s\n", argv[3]);
      exit(-1);
   }

   while (len > BUF_SIZE)
   {
      fread(buf, BUF_SIZE, 1, fp);
      fwrite(buf, BUF_SIZE, 1, fout);
      len -= BUF_SIZE;
   }

   fread(buf, len, 1, fp);
   fwrite(buf, len, 1, fout);

   fclose(fp);
   fclose(fout);

   return 0;
}



u9bmp.c
/*

Ultima 9 Bitmap Viewer
Copyright (C) 2000 David Churchill

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

The author can be contacted by electronic mail at froods@alphalink.com.au
The most recent version can be found at http://home.alphalink.com.au/~froods/

Ultima 9 is probably a registered trademark of Origin Systems, Inc.
The author of this program is in no way affiliated with Origin Systems.

*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <conio.h>
#include <errno.h>
#include <ctype.h>
#include <allegro.h>
#include "flxx.h"


#define BITMAP_NAME "bitmap16.flx"
#define NUM_MODES 3


struct bmp16hdr_struct
{
   char  id1, id2;
   int   filesize;
   int   res,       /* reserved */
         dataofs,
         hdrsize,
         width,
         height;
   short planes,
         bpp;
   int   comp,
         datasize,
         hres,
         vres,
         colors,
         impcolors,
         mask_red,
         mask_green,
         mask_blue;
} __attribute__ ((packed));


struct bmp16hdr_struct base_hdr =
{ 'B', 'M', 0, 0, 0x42, 0x28, 0, 0, 1, 16, 3, 0, 0, 0, 0, 0,
   0xF800, 0x07E0, 0x001F };
/* base_hdr contains the correct values for a 16-bit colour
   bitmap.  fields that need to be filled in are filesize, width,
   height, datasize. */



void exit_err(char *fmt, char *param)
{
   set_gfx_mode(GFX_TEXT, 80, 25, 0, 0);
   fprintf(stderr, fmt, param);
   exit(-1);
}



int write_bmp16(char *name, char *img, int width, int height)
{
   FILE *fo;
   int padlen = 0, i;
   char padding[4] = {'\0', '\0', '\0', '\0'};
   struct bmp16hdr_struct bmphdr;


   fo = fopen(name, "rb");
   if (fo != NULL)
   {
      fclose(fo);
      return -EEXIST;
   }

   fo = fopen(name, "wb");
   if (fo == NULL)
      return -errno;


   if ((width*2) % 4)                /* each scanline must end on a 4-byte */
      padlen = 4 - ((width*2) % 4);  /* boundary */

   memcpy(&bmphdr, &base_hdr, sizeof(struct bmp16hdr_struct));
   bmphdr.width = width;
   bmphdr.height = height;
   bmphdr.datasize = height * (width * 2 + padlen);
   bmphdr.filesize = sizeof(struct bmp16hdr_struct) + bmphdr.datasize;

   fwrite(&bmphdr, sizeof(struct bmp16hdr_struct), 1, fo);

   img += (height - 1) * width * 2;

   for (i=0; i<height; i++)
   {
      fwrite(img, width * 2, 1, fo);
      if (padlen)
         fwrite(padding, padlen, 1, fo);
      img -= width * 2;
   }


   fclose(fo);
   return 0;
}



void show_texture(FILE *fp, int id, unsigned short mask, int dump)
{
   unsigned short *img;
   int width, height, frame, maxframe;
   unsigned int ofs, len;
   int x = 0, y = 0, i, j, col;
   char fname[80];


   if (flxx_getitem(fp, id) <= 0)
      return;

   fseek(fp, 0x08, SEEK_CUR);
   fread(&maxframe, 4, 1, fp);

   for (frame=0; frame<maxframe; frame++)
   {
      flxx_getitem(fp, id);
      fseek(fp, 0x10 + frame * 8, SEEK_CUR);
      fread(&ofs, 4, 1, fp);
      fread(&len, 4, 1, fp);
      fseek(fp, ofs - 0x18 - frame * 8, SEEK_CUR);

      fseek(fp, 4, SEEK_CUR);
      fread(&width, 4, 1, fp);
      if ((frame != 0) && (x + width > SCREEN_W))
      {
         x = 0;
         y += height + 2;
      }
      fread(&height, 4, 1, fp);

      img = malloc(width * height * 2);
      if (img == NULL)
         exit_err("out of memory%s\n", "");

      fseek(fp, 8 + 4 * height, SEEK_CUR);
      fread(img, 1, width * height * 2, fp);

      for (j=0; j<height; j++)
         for (i=0; i<width; i++)
         {
            col = img[j*width+i] & mask;
            putpixel(screen, x+i, y+j, col);
         }

      if (dump)
      {
         for (i=0; i<height*width; i++)
            img[i] &= mask;
         sprintf(fname, "u9-%04i-%03i.bmp", id, frame);
         write_bmp16(fname, (char *)img, width, height);
      }


      free(img);

      x += width + 2;
   }
}



int main(int argc, char *argv[])
{
   FILE *fp;
   int curritem = 0, numitems, i;
   unsigned short mask = 0xFFFF;
   char ch = '\0';
   int done = 0,
       wind = 0,          /* not as in air */
       mode = 0,
       refresh = 1;

   struct gfx_modes_struct { int w,h; } gfx_modes[NUM_MODES] =
      { {640, 480}, {800, 600}, {1024, 768} };


   if (argc != 0)
   {
      if ( (strchr(argv[1], '?') != NULL) ||
           (strchr(argv[1], 'h') != NULL) ||
           (strchr(argv[1], 'H') != NULL) )
      {
         fprintf(stderr, "usage: %s [gfx-mode-id]\n", argv[0]);
         fprintf(stderr, "gfx-mode-id is one of:\n");
         for (i=0; i<NUM_MODES; i++)
            fprintf(stderr, "  %i => %ix%ix16 %s\n",
                    i, gfx_modes[i].w, gfx_modes[i].h,
                    (i == 0) ? "(default)" : "");
         exit(-1);
      }

      mode = atoi(argv[1]);
      if ((mode < 0) || (mode >= NUM_MODES))
         mode = 0;
   }


   fp = flxx_open(BITMAP_NAME);
   if (fp == NULL)
      exit_err("error opening %s\n", BITMAP_NAME);
   numitems = flxx_numitems(fp);


   allegro_init();
   set_color_depth(16);
   if (set_gfx_mode(GFX_AUTODETECT, gfx_modes[mode].w, gfx_modes[mode].h,
       0, 0) < 0)
      exit_err("error initialising graphics:\n   %s\n", allegro_error);


   while (flxx_getitem(fp, curritem) <= 0)
   {
      curritem++;
      if (curritem > numitems)
         exit_err("no non-zero items found!!\n", "");
   }



   while (!done)
   {
      if (refresh)
      {
         clear(screen);
         show_texture(fp, curritem, mask, 0);
         textprintf(screen, font, 10, SCREEN_H - 20, 0xFFFF,
                    "img %i/%i | mask 0x%X", curritem, numitems-1, mask);
         refresh = 0;
      }

      ch = toupper(getch());
      switch (ch)
      {
         case 0x1B:  done = 1;
                     break;

         case 'V':   wind =  1;       break;
         case 'F':   wind = -1;       break;
         case 'C':   wind =  10;      break;
         case 'D':   wind = -10;      break;
         case 'X':   wind =  100;     break;
         case 'S':   wind = -100;     break;
         case 'Z':   wind =  1000;    break;
         case 'A':   wind = -1000;    break;

         case 0x0D:  clear(screen);
                     show_texture(fp, curritem, mask, 1);
                     textprintf(screen, font, 10, SCREEN_H - 20, 0xFFFF,
                                "dumped.");
                     break;

         case ' ':   mask ^= 0x8400;  /* flip b/w 0xffff and 0x7bff */
                     refresh = 1;
                     break;
      }

      if (wind != 0)
      {
         curritem += wind;

         if (curritem < 0)
            curritem += numitems;
         if (curritem >= numitems)
            curritem -= numitems;

         while (flxx_getitem(fp, curritem) <= 0)
         {
            if (wind < 0)
               curritem--;
            else
               curritem++;

            if (curritem < 0)
               curritem = numitems - 1;
            if (curritem >= numitems)
               curritem = 0;
         }

         wind = 0;
         refresh = 1;
      }

   }

   fclose(fp);

   return 0;
}



u9txv.c
/*

Ultima 9 Texture Viewer
Copyright (C) 2000 David Churchill

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

The author can be contacted by electronic mail at froods@alphalink.com.au
The most recent version can be found at http://home.alphalink.com.au/~froods/

Ultima 9 is probably a registered trademark of Origin Systems, Inc.
The author of this program is in no way affiliated with Origin Systems.

*/


#include <stdio.h>
#include <stdlib.h>
#include <conio.h>
#include <allegro.h>
#include "flxx.h"



void exit_err(char *fmt, char *param)
{
   set_gfx_mode(GFX_TEXT, 80, 25, 0, 0);
   fprintf(stderr, fmt, param);
   exit(-1);
}



int *load_palette(void)
{
   int *pal;
   FILE *fpal;



   fpal = fopen("ankh.pal", "rb");
   if (fpal == NULL)
      exit_err("error opening ankh.pal%s\n", "");

   if ( ( pal = malloc(256*sizeof(int)) ) == NULL)
      exit_err("no memory for pallete!%s\n", "");

   fread(pal, 4, 256, fpal);
   fclose(fpal);

   return pal;
}



void show_texture(FILE *fp, int *pal, int id)
{
   unsigned char *img;
   int width, height, len;
   static int x = 0, y = 0;
   int i, j, col;



/*   clear(screen); */

   len = flxx_getitem(fp, id);
   if (len <= 0)
      return;

   fread(&width, 4, 1, fp);
   fread(&height, 4, 1, fp);
   img = malloc(width * height);
   if (img == NULL)
      exit_err("out of memory%s\n", "");

   fseek(fp, len - width * height - 8, SEEK_CUR);
   fread(img, 1, width * height, fp);

   for (i=0; i<height; i++)
      for (j=0; j<width; j++)
      {
         col = pal[img[j*width+i]];
         putpixel(screen, x+i,   y+j,   col);
      }


   x += width;
   if ((x+width) > SCREEN_W)
   {
      x = 0;
      y += height;
      if ((y+height) > SCREEN_H)
         y = 0;
   }

   free(img);
}



int main(int argc, char *argv[])
{
   FILE *fp;
   int *pal;
   int curritem = 0, numitems;
   char ch = '\0';
   int done = 0;


   allegro_init();


   if (argc != 2)
      exit_err("usage: %s <texturefile>\n", argv[0]);


   fp = flxx_open(argv[1]);
   if (fp == NULL)
      exit_err("error opening %s\n", argv[1]);
   numitems = flxx_numitems(fp);


   set_color_depth(24);
   if (set_gfx_mode(GFX_AUTODETECT, 1024, 768, 0, 0) < 0)
      exit_err("error initialising graphics:\n   %s\n", allegro_error);
   pal = load_palette();



   while (flxx_getitem(fp, curritem) == 0)
   {
      curritem++;
      if (curritem > numitems)
         exit_err("no non-zero items found!\n", "");
   }

   while (!done)
   {
      show_texture(fp, pal, curritem);
      ch = getch();
      switch (ch)
      {
         case 0x1B:  done = 1;
                     break;

         case ',':   curritem--;
                     if (curritem < 0)
                        curritem = numitems - 1;
                     while (flxx_getitem(fp, curritem) == 0)
                     {
                        curritem--;
                        if (curritem < 0)
                           curritem = numitems - 1;
                     }
                     break;

         case '.':   curritem++;
                     if (curritem >= numitems)
                        curritem = 0;
                     while (flxx_getitem(fp, curritem) == 0)
                     {
                        curritem++;
                        if (curritem >= numitems)
                           curritem = 0;
                     }
                     break;
      }
   }

   free(pal);
   fclose(fp);

   return 0;
}



u9bmp-linux.c
/*

Ultima 9 Bitmap Viewer
Copyright (C) 2000 David Churchill

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

The author can be contacted by electronic mail at froods@alphalink.com.au
The most recent version can be found at http://home.alphalink.com.au/~froods/

Ultima 9 is probably a registered trademark of Origin Systems, Inc.
The author of this program is in no way affiliated with Origin Systems.

*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include "flxx.h"

#include "SDL.h"

#define SCREEN_W 800
#define SCREEN_H 600


#define BITMAP_NAME "bitmap16.flx"
#define NUM_MODES 3

SDL_Surface *screen = NULL;


struct bmp16hdr_struct
{
   char  id1, id2;
   int   filesize;
   int   res,       /* reserved */
         dataofs,
         hdrsize,
         width,
         height;
   short planes,
         bpp;
   int   comp,
         datasize,
         hres,
         vres,
         colors,
         impcolors,
         mask_red,
         mask_green,
         mask_blue;
} __attribute__ ((packed));


struct bmp16hdr_struct base_hdr =
{ 'B', 'M', 0, 0, 0x42, 0x28, 0, 0, 1, 16, 3, 0, 0, 0, 0, 0,
   0xF800, 0x07E0, 0x001F };
/* base_hdr contains the correct values for a 16-bit colour
   bitmap.  fields that need to be filled in are filesize, width,
   height, datasize. */

SDLKey getch(void) {
  SDL_Event event;
  SDLKey key;
  char quit = 0;

  while( !quit && SDL_WaitEvent( &event ) ){
    switch( event.type ){
	  case SDL_QUIT:
		key = SDLK_ESCAPE;
		quit = 1;
		break;
      case SDL_KEYUP:
		key = event.key.keysym.sym;
        quit = 1;
        break;

      default:
        break;
    }
  }

  return key;
}


/* Taken from http://sdl.beuc.net/sdl.wiki/Pixel_Access */
void putpixel(SDL_Surface *surface, int x, int y, Uint32 pixel)
{
    int bpp = surface->format->BytesPerPixel;
    /* Here p is the address to the pixel we want to set */
    Uint8 *p = (Uint8 *)surface->pixels + y * surface->pitch + x * bpp;

	if(x < 0) return;
	if(y < 0) return;
	if(x > SCREEN_W - 1) return;
	if(y > SCREEN_H - 1) return;

    switch(bpp) {
    case 1:
        *p = pixel;
        break;

    case 2:
        *(Uint16 *)p = pixel;
        break;

    case 3:
        if(SDL_BYTEORDER == SDL_BIG_ENDIAN) {
            p[0] = (pixel >> 16) & 0xff;
            p[1] = (pixel >> 8) & 0xff;
            p[2] = pixel & 0xff;
        } else {
            p[0] = pixel & 0xff;
            p[1] = (pixel >> 8) & 0xff;
            p[2] = (pixel >> 16) & 0xff;
        }
        break;

    case 4:
        *(Uint32 *)p = pixel;
        break;
    }
}

void clear(SDL_Surface *surface) {
   SDL_Rect rect;

   rect.x = 0;
   rect.y = 0;
   rect.w = surface->w;
   rect.h = surface->h;
   SDL_FillRect(surface, &rect, SDL_MapRGB(surface->format, 0, 0, 0));
}


void exit_err(char *fmt, char *param)
{
   fprintf(stderr, fmt, param);
   exit(-1);
}


int write_bmp16(char *name, char *img, int width, int height)
{
   FILE *fo;
   int padlen = 0, i;
   char padding[4] = {'\0', '\0', '\0', '\0'};
   struct bmp16hdr_struct bmphdr;


   fo = fopen(name, "rb");
   if (fo != NULL)
   {
      fclose(fo);
      return -EEXIST;
   }

   fo = fopen(name, "wb");
   if (fo == NULL)
      return -errno;


   if ((width*2) % 4)                /* each scanline must end on a 4-byte */
      padlen = 4 - ((width*2) % 4);  /* boundary */

   memcpy(&bmphdr, &base_hdr, sizeof(struct bmp16hdr_struct));
   bmphdr.width = width;
   bmphdr.height = height;
   bmphdr.datasize = height * (width * 2 + padlen);
   bmphdr.filesize = sizeof(struct bmp16hdr_struct) + bmphdr.datasize;

   fwrite(&bmphdr, sizeof(struct bmp16hdr_struct), 1, fo);

   img += (height - 1) * width * 2;

   for (i=0; i<height; i++)
   {
      fwrite(img, width * 2, 1, fo);
      if (padlen)
         fwrite(padding, padlen, 1, fo);
      img -= width * 2;
   }


   fclose(fo);
   return 0;
}



void show_texture(FILE *fp, int id, unsigned short mask, int dump)
{
   unsigned short *img = NULL;
   int width, height, frame, maxframe;
   unsigned int ofs, len;
   int x = 0, y = 0, i, j, col;
   char fname[80] = {0};

   if (flxx_getitem(fp, id) <= 0)
      return;

   fseek(fp, 0x08, SEEK_CUR);
   fread(&maxframe, 4, 1, fp);

   fprintf(stdout, "show_texture: item has %d frames\n", maxframe);

   for (frame=0; frame<maxframe; frame++)
   {
      flxx_getitem(fp, id);

      fseek(fp, 0x10 + frame * 8, SEEK_CUR);
	  fprintf(stdout, "show_texture: frame %d, seeked to 0x%x (%ld)\n", frame, (unsigned int)ftell(fp), ftell(fp));
      fread(&ofs, 4, 1, fp);
      fread(&len, 4, 1, fp);
      fseek(fp, ofs - 0x18 - frame * 8, SEEK_CUR);

      fseek(fp, 4, SEEK_CUR);
      fread(&width, 4, 1, fp);
      if ((frame != 0) && (x + width > SCREEN_W))
      {
         x = 0;
         y += height + 2;
      }
      fread(&height, 4, 1, fp);

      img = malloc(width * height * 2);
      if (img == NULL) {
         exit_err("out of memory%s\n", "");
      }

      fseek(fp, 8 + 4 * height, SEEK_CUR);
      fread(img, 1, width * height * 2, fp);

	SDL_LockSurface(screen);
      for (j=0; j<height; j++)
         for (i=0; i<width; i++)
         {
            col = img[j*width+i] & mask;
            putpixel(screen, x+i, y+j, col);
         }
	SDL_UnlockSurface(screen);

      if (dump)
      {
         for (i=0; i<height*width; i++)
            img[i] &= mask;
         sprintf(fname, "u9-%04i-%03i.bmp", id, frame);
         write_bmp16(fname, (char *)img, width, height);
      }

      free(img);

      x += width + 2;
   }

   SDL_Flip(screen);

	fprintf(stdout, "show_texture: finished.\n");
}



int main(int argc, char *argv[])
{
   FILE *fp = NULL;
   int curritem = 0, numitems;
   unsigned short mask = 0xFFFF;
   SDLKey key;
   int done = 0,
       wind = 0,          /* not as in air */
       refresh = 1;

   fprintf(stdout, "Opening BITMAP_NAME...\n");
   fp = flxx_open(BITMAP_NAME);
   if (fp == NULL) {
      exit_err("error opening %s\n", BITMAP_NAME);
   } else {
      fprintf(stdout, "BITMAP_NAME opened successfully.\n");
   }
   numitems = flxx_numitems(fp);


   if(SDL_Init(SDL_INIT_VIDEO) < 0) {
      fprintf(stderr, "Could not initialize SDL: %s\n", SDL_GetError());
      exit(-1);
   } else {
      fprintf(stdout, "SDL initialized successfully.\n");
   }

   screen = SDL_SetVideoMode(SCREEN_W, SCREEN_H, 16, 0);
   if(screen == NULL) {
      fprintf(stderr, "Could not set video mode: %s\n", SDL_GetError());
      exit(-1);
   }
   //allegro_init();
   //set_color_depth(16);
   //if (set_gfx_mode(GFX_AUTODETECT, gfx_modes[mode].w, gfx_modes[mode].h,
   //    0, 0) < 0)
   //   exit_err("error initialising graphics:\n   %s\n", allegro_error);


   while (flxx_getitem(fp, curritem) <= 0)
   {
      curritem++;
      if (curritem > numitems)
         exit_err("no non-zero items found!!\n", "");
   }



   while (!done)
   {
      if (refresh)
      {
         clear(screen);
         show_texture(fp, curritem, mask, 0);
         //textprintf(screen, font, 10, SCREEN_H - 20, 0xFFFF,
         //           "img %i/%i | mask 0x%X", curritem, numitems-1, mask);
         fprintf(stdout, "img %i/%i | mask 0x%X\n", curritem, numitems-1, mask);
		fflush(stdout);
         refresh = 0;
      }

      key = getch();
      switch (key)
      {
		// quit program
         case SDLK_ESCAPE:
			done = 1;
            break;

		// cycle through the images
         case SDLK_RIGHT:   wind =  1;       break;
         case SDLK_LEFT:   wind = -1;       break;
         case SDLK_c:   wind =  10;      break;
         case SDLK_d:   wind = -10;      break;
         case SDLK_x:   wind =  100;     break;
         case SDLK_s:   wind = -100;     break;
         case SDLK_z:   wind =  1000;    break;
         case SDLK_a:   wind = -1000;    break;

			// hit 'f' to dump
         case SDLK_f:  clear(screen);
                     show_texture(fp, curritem, mask, 1);
                     //textprintf(screen, font, 10, SCREEN_H - 20, 0xFFFF,
                     //           "dumped.");
                     fprintf(stdout, "dumped.\n");
                     break;

		// flip the mask
         case SDLK_SPACE:   mask ^= 0x8400;  /* flip b/w 0xffff and 0x7bff */
                     refresh = 1;
                     break;
		 default:
			break;
      }

      if (wind != 0)
      {
         curritem += wind;

         if (curritem < 0)
            curritem += numitems;
         if (curritem >= numitems)
            curritem -= numitems;

		 fprintf(stdout, "looking for item %d\n", curritem);

         while (flxx_getitem(fp, curritem) <= 0)
         {
            if (wind < 0)
               curritem--;
            else
               curritem++;

            if (curritem < 0)
               curritem = numitems - 1;
            if (curritem >= numitems)
               curritem = 0;
         }

		fprintf(stdout, "ending up on item %d (assuming junk inbetween?)\n", curritem);

         wind = 0;
         refresh = 1;
      }

   }

   fclose(fp);

   SDL_Quit();

   return 0;
}

